src/eddymotion/model/kernels.py

```diff
@@ -0,0 +1,186 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY kIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# We support and encourage derived works from this project, please read
+# about our expectations at
+#
+#     https://www.nipreps.org/community/licensing/
+#
+import numpy as np
+from sklearn.gaussian_process.kernels import Hyperparameter, Kernel
+
+
+class SphericalCovarianceKernel(Kernel):
+    """
+    Custom kernel based on spherical covariance function.
+
+    Parameters
+    ----------
+    lambda_s : :obj:`float`
+        Scale parameter for the covariance function.
+    a : :obj:`float`
+        Distance parameter where the covariance function goes to zero.
+    sigma_sq : :obj:`float`
+        Noise variance term.
+    lambda_s_bounds : :obj:`tuple` of :obj:`float` or "fixed"
+        The lower and upper bound on lambda_s.
+    a_bounds : :obj:`tuple` of :obj:`float` or "fixed"
+        The lower and upper bound on a.
+    sigma_sq_bounds : :obj:`tuple` of :obj:`float` or "fixed"
+        The lower and upper bound on sigma_sq.
+    """
+
+    def __init__(
+        self,
+        lambda_s=2.0,
+        a=0.1,
+        sigma_sq=1.0,
+        lambda_s_bounds=(1e-5, 1e4),
+        a_bounds=(1e-5, np.pi),
+        sigma_sq_bounds=(1e-5, 1e4),
+    ):
+        self.lambda_s = lambda_s
+        self.a = a
+        self.sigma_sq = sigma_sq
+        self.lambda_s_bounds = lambda_s_bounds
+        self.a_bounds = a_bounds
+        self.sigma_sq_bounds = sigma_sq_bounds
+
+    @property
+    def hyperparameter_lambda_s(self):
+        return Hyperparameter("lambda_s", "numeric", self.lambda_s_bounds)
+
+    @property
+    def hyperparameter_a(self):
+        return Hyperparameter("a", "numeric", self.a_bounds)
+
+    @property
+    def hyperparameter_sigma_sq(self):
+        return Hyperparameter("sigma_sq", "numeric", self.sigma_sq_bounds)
+
+    def __call__(self, X, Y=None, eval_gradient=False):
+        """
+        Compute the kernel matrix.
+
+        Parameters
+        ----------
+        X : :obj:`array-like` of shape (n_samples, n_samples)
+            Precomputed pairwise angles.
+        Y : :obj:`array-like` of shape (n_samples, n_samples), optional
+            Second input for the kernel function.
+        eval_gradient : :obj:`bool`
+            Evaluate the gradient with respect to the log of
+            the kernel hyperparameter.
+
+        Returns
+        -------
+        K : :obj:`array-like` of shape (n_samples, n_samples)
+            Kernel matrix.
+        K_gradient : :obj:`array-like` of shape (n_samples, n_samples, n_dims), optional
+            The gradient of the kernel matrix with respect to the log of the
+            hyperparameters. Only returned when `eval_gradient` is True.
+        """
+        if Y is not None:
+            X = Y
```

Comment:
So I see that in https://scikit-learn.org/stable/modules/generated/sklearn.gaussian_process.kernels.Kernel.html they do the `X=Y` assignment if `Y is None` (note the absence of the `not`).

What if we avoid doing the `X=Y` assignment to get past this situation? My thought is linked to 
This is related to https://github.com/nipreps/eddymotion/pull/203/files#r1641440923, and I now completely see your point in https://github.com/nipreps/eddymotion/pull/200#discussion_r1615824997 @esavary:

```
if Y is None:
  theta = compute_pairwise_angles(X, X)
else:
  theta = compute_pairwise_angles(X, Y)

 K = self.lambda_s * np.where(
                theta <= self.a, 1 - 3 * (theta / self.a) ** 2 + 2 * (theta / self.a) ** 3, 0
            )
```

(assuming here `X` and potentially `Y` are a set of gradient vectors provided by the caller.)

(Have not checked whether we would still need to multiply by `np.eye(len(theta)` if `Y is None`)

The `eval_gradient` block may need to be adjusted concerning the dimensions now that `X` is assumed to be `bvecs`.

So `compute_pairwise_angles` would need to be modified to accept two input parameters. I apologize for not having realized about the downstream needs and implications of these two aspects. Really sorry @esavary.

Reply:
> So `compute_pairwise_angles` would need to be modified to accept two input parameters. I apologize for not having realized about the downstream needs and implications of these two aspects. Really sorry @esavary.
No worries, it also makes a lot of sense to compute the angles beforehand since we have all the information at the beginning. It would just have required us to recode the fit method. I'll suggest a generalization of the angle computation in a separate PR and then use this to update the kernel to take the bvec directly. This will also solve the problem of X=Y because we can then perform different computations when Y is given.

Reply:
PR #212 to make `compute_pairwise_angles` accept two parameters.